Dem insertion: clear and add (#1135)
chaos-polymtl/lethe: doc/source/parameters/dem/insertion_info.rst

.. code-block:: text

  subsection insertion info

>     # Choices are volume|plane|list|file|clear_and_add

Reviewer: what do you think of "empty_and_fill", "empty_and_refill", or "dump_and_fill"/"dump_and_refill"?

Reviewer 2: I would suggest remove_and_insert or erase_and_insert
Clearing gives us the idea of clearing the memory of a data structure. In essence, you are really removing particles.
I really prefer to use the words "remove" and "insert" because these also correspond to the words used in the ParticleHandler class.

    # Every method
    set insertion frequency                            = 20000

>     set clearing particles                             = false

Reviewer 2: I'm not sure I like the name clearing. The deal.II nomenclature is "remove" for particles. I would rather we remain coherent.

    # If method = file
    set insertion file name                            = particles.input

    # If method = clear_and_add
    set list of input files                            = particles.input

>     set clearing box points coordinates                = 0., 0., 0.: 1., 1., 1.

Reviewer: This falls in the problem of not being suitable for spherical and cylindrical domains. Could you do a mesh instead of a box? Then it could be a deal.ii mesh or a custom mesh.

Author: I like the idea, but currently I don't know how to do that and it won't be useful for my problem. I prefer not to put to much time into something that will probably never be used.

I think this would be a good addition long term, especially if this addition is seamless with the current implementation.

Author: Also, we could do the same thing for the volume insertion.

Reviewer 2: This aspect is orthogonal. Here the idea is that you have a clearing/insertion mechanism. The fact that we don't have another box than "square box" is an another issue that we can address in another PR. One PR, one feature.

    set list diameters                                 = 0.

    # If method = file

>     set list of input files                            = particles.input

Reviewer: input file list

* The ``insertion maximum offset`` and ``insertion prn seed`` parameters defines the random offset values to the initial positions of particles during a ``volume`` and ``plane`` insertion. The ``insertion maximum offset`` parameter defines the maximum value for an offset. The ``insertion prn seed`` parameter defines the pseudo-random number (PRN) with which offset values are getting generated.

> * The ``clearing box points coordinates`` parameter defines a clearing box where particles will be removed from the triangulation just before the insertion of new particles. This assures that there is no excessive overlap between new and previously inserted particles. If the ``clearing particles`` parameter is set to ``false`` particle removal will not be performed.

Reviewer: "box vertices" instead of "box points coordinates"?

Author: This is to remain coherent with the insertion box of the volume insertion.

---------------------
File
---------------------

> The ``file`` insertion method insert particles in a similar way to the ``list`` insertion method. The main difference between these two methods is the option to use external files provided by the ``list of input files`` parameter. This parameter is set at ``particles.input`` by default, but a list a files can be specified. At each insertion time step, a different file will be used. If the end of the list is reached and there are still particles to be inserted, the list returns to the first file. An insertion file must follow this structure:

Reviewer: If you change the name of the parameters, don't forget to change them here :)

Author: Good catch

> The ``file`` insertion method insert particles in a similar way to the ``list`` insertion method. The main difference between these two methods is the option to use external files provided by the ``list of input files`` parameter. This parameter is set at ``particles.input`` by default, but a list a files can be specified. At each insertion time step, a different file will be used. If the end of the list is reached and there are still particles to be inserted, the list returns to the first file. An insertion file must follow this structure:

Reviewer: ```suggestion
The ``file`` insertion method inserts particles in a similar way to the ``list`` insertion method. The main difference between these two methods is the option to use external files provided by the ``list of input files`` parameter. This parameter is set at ``particles.input`` by default, but a file list can be specified. At each insertion time step, a different file will be used. If the end of the list is reached and there are still particles to be inserted, the list returns to the first file. An insertion file must follow this structure:
```
```suggestion
The ``file`` insertion method insert particles in a similar way to the ``list`` insertion method. The main difference between these two methods is the option to use external files provided by the ``list of input files`` parameter. This parameter is set at ``particles.input`` by default, but a list a files can be specified. At each insertion time step, a different file will be used. If the end of the list is reached and there are still particles to be inserted, the list returns to the first file. An insertion file must follow this structure:
```

    The ``file`` insertion combine with the ``extract-particles-properties-from-vtu.py`` python code can be a useful tool. The loading of particles and the rest of the simulation can be performed in two different triangulations, witch is not the case of the the restart feature. This means that the loading triangulation can have smaller cells and a bigger domain to allow for the use of larger insertion boxes. Then, particles properties can be extracted and the remainder of the simulation can be performed in the appropriate triangulation.

.. warning::

>     The critical Rayleigh time step is computed from the parameters in the ``particle type`` subsections, not the ``insertion info`` subsection. It is the user's responsibility to fill the ``particle type`` subsections correctly according to the diameter values stored in the insertion input file, otherwise Rayleigh time percentage displayed at the start of every DEM simulation may not be accurate.

Reviewer: What was changed here?

Author: Idk... 🤷🏼‍♂️

.. note::

>     The ``clear_and_add`` insertion method shouldn't be used for most application. It can be extremely useful for really specific DEM simulation, but remain an experimental feature.

Reviewer: ```suggestion
    The ``clear_and_add`` insertion method is not suited to most applications. It can be useful for some DEM simulations, but this remains an experimental feature.
```

Reviewer: What applications, for example? I would add examples here, since you don't have an example of use in the documentation.

Reviewer: Additionally, it is important to mention that the feature, as is, is limited to a hexahedron.
